music_this_week_app/templates/music_this_week/setup.html

```diff
@@ -12,23 +12,75 @@
         <p><a href="{{ profile }}" target="_blank">{{ username }}</a>&nbsp; &nbsp;&nbsp;<a href="/login/?retry=true">Switch Users</a></p>
         <p><span style="text-decoration: underline;">Search Arguments:</span></p>
         <form action="/search" id="form">
-            <p><strong>Location:</strong> <input id = "loc_field" type="text" name="location" value="San Francisco"></p>
+            <p id = 'location_options'><strong>Location:</strong> <input id = "loc_field" type="text" name="location" value="{{location}}"> 
+               <input type = "checkbox" id = "latlng_checkbox" name = "latlng_checkbox" onchange = "clickCheckBox.call(this)"> Use my current location </input></p>
             <p><strong>Time range:</strong> <input id = "date_field" type="text" name="date" value="Next 7 Days"></p>
+            <p><input id = "hidden_lat" hidden = true name = "hidden_lat"></p>
+            <p><input id = "hidden_lon" hidden = true name = "hidden_lon"></p>
             <input hidden=true name="nResults" value="200">
             <button type="submit" id="search-button" onclick="buttonClick(this.form)">Search</button>
         </form>
         <p>&nbsp;</p>
         <p id="wait-text" hidden="True">Please be patient. In few minutes a playlist called musicThisWeek will appear in your spotify app.<br><br>If you close this page, the playlist will still be created in the background.</p>
     </div>
-
     <script>
+        //init a variable that will be true if checkbox clicked and user approves location request
+        var userLocationProvided = false;
         //When Search button is pressed, change the UI and then submit the request.
         function buttonClick(form) {
             document.getElementById('search-button').innerHTML = "Searching...";
             document.getElementById('search-button').disabled = "True";
             document.getElementById('wait-text').hidden = false;
             form.submit();
         }
+        //When checkbox is clicked request users location if we havent already
+        function clickCheckBox(){
```

Comment:
There's some complicated logic in this function to manage checkbox and text field state that I don't quite understand. Can you please add a detailed comment explaining how it all works? 

One thing that _may_ clear things up, but I'm not sure, is to change (and rename) the `toggleLocationField` function to explicitly set the state to Hidden or Shown (or True or False or Checked or Unchecked or whatever) rather than just changing it from whatever it was last. It seems to work now, but might be more readable that way. (not entirely sure about this).

I'm slightly confused about where the loc_field (which I guess refers to the text box?) is hidden and shown. It can be toggle True or False by the `toggleLocationField` function, but also set to hidden by the `locationProvided` callback.